memtable/write_buffer_manager.cc

```diff
@@ -128,6 +173,45 @@ void WriteBufferManager::FreeMemWithCache(size_t mem) {
 #endif  // ROCKSDB_LITE
 }
 
+void WriteBufferManager::MaybeFlushLocked(DB* this_db) {
+  if (!ShouldFlush()) {
+    return;
+  }
+  // We only flush at most one column family at a time.
+  // This is enough to keep size under control except when flush_size is
+  // dynamically decreased. That case is managed in `SetFlushSize`.
+  WriteBufferSentinel* candidate = nullptr;
+  uint64_t candidate_size = 0;
+  uint64_t max_score = 0;
+  uint64_t current_score = 0;
+  for (auto& s : sentinels_) {
+    uint64_t current_memory_bytes = std::numeric_limits<uint64_t>::max();
+    uint64_t oldest_time = std::numeric_limits<uint64_t>::max();
+    s->db->GetApproximateActiveMemTableStats(s->cf, &current_memory_bytes,
+                                             &oldest_time);
+    if (flush_oldest_first_) {
+      // Convert oldest to highest score.
+      current_score = std::numeric_limits<uint64_t>::max() - oldest_time;
+    } else {
+      current_score = current_memory_bytes;
+    }
+    if (current_score > max_score) {
+      candidate = s.get();
+      max_score = current_score;
+      candidate_size = current_memory_bytes;
+    }
+  }
+
+  if (candidate != nullptr) {
+    FlushOptions flush_opts;
+    flush_opts.allow_write_stall = true;
+    flush_opts.wait = false;
+    flush_opts._write_stopped = (candidate->db == this_db);
+    flush_opts.min_size_to_flush = candidate_size - 1;
+    candidate->db->Flush(flush_opts, candidate->cf);
```

Comment:
As discussed offline, it should not be flushed directly. Instead it should call a hook.

Reply:
Unless the candidate is the exact the db that tries to write, then it's OK to flush directly.

Reply:
The code depends on `MaybeFlush` releasing memory immediately (via `SwitchMemtable`). So it cannot be replaced with hooks that don't immediately call flush. My plan is to add a `manual` flag to it, when `manual==true`, user is responsible for calling `MaybeFlush` instead of RocksDB write thread.